documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Reverses the sign of each component of the quaternion.</summary>
+    /// <param name="value">The quaternion to negate.</param>
+    /// <returns>The negated quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Negate<T>(Quaternion<T> value)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Divides each component of a specified <see cref="Quaternion{T}" /> by its length.</summary>
+    /// <param name="value">The quaternion to normalize.</param>
+    /// <returns>The normalized quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Normalize<T>(Quaternion<T> value)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Interpolates between two quaternions, using spherical linear interpolation.</summary>
+    /// <param name="quaternion1">The first quaternion.</param>
+    /// <param name="quaternion2">The second quaternion.</param>
+    /// <param name="amount">The relative weight of the second quaternion in the interpolation.</param>
+    /// <returns>The interpolated quaternion.</returns>
+    public static Quaternion<T> Slerp<T>(Quaternion<T> quaternion1, Quaternion<T> quaternion2, T amount)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>, IComparisonOperators<T, T, bool>;
+
+    /// <summary>Subtracts each element in a second quaternion from its corresponding element in a first quaternion.</summary>
+    /// <param name="value1">The first quaternion.</param>
+    /// <param name="value2">The second quaternion.</param>
+    /// <returns>The quaternion containing the values that result from subtracting each element in <paramref name="value2" /> from its corresponding element in <paramref name="value1" />.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Subtract<T>(Quaternion<T> value1, Quaternion<T> value2)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+    
+    /// <summary>Calculates the length of the quaternion.</summary>
+    /// <returns>The computed length of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T Length<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Calculates the squared length of the quaternion.</summary>
+    /// <returns>The length squared of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T LengthSquared<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+}
+```
+
+# Geometric Types
+
+The following Geometric Types are defined:
+- Box2DF
+- Box2DI
+- Box3DF
+- Box3DI
+- CircleF
+- CircleI
+- CubeF
+- CubeI
+- PlaneF
+- PlaneI
+- Ray2DF
+- Ray2DI
+- Ray3DF
+- Ray3DI
```

Comment:
Drop the D, as we're already doing so for vectors. Possibly rename Box to Aabb or AxisAlignedBoundingBox, as the current name is rather confusing as we have both Box and Rectangle.

Reply:
I did remove the D, but Box2D and Rectangle are 2 separate representations of the same object. If they sound similar that's because they are.

Reply:
I’d like to keep Box2I and Box2F named as such, not calling it AABB was explicit on my part (for no particular reason I just don’t like it)

Agree with renaming Ray3DI to just Ray3I as well, the Ds were for consistency and that’s no longer the case here.